File: 22-generate-parentheses/generate-parentheses.py
class Solution(object):
    def generateParenthesis(self, n):
        """
        :type n: int
        :rtype: List[str]
        """
        slots = n * 2
        # Recursive
        def generate(sofar = [""]):
            if len(sofar[0]) == slots:
                return sofar

            L = []
            for paren in sofar:
                L.append(paren + "(")
                L.append(paren + ")")
            return generate(L)

        L = generate()

        def is_valid(paren):
            op = 0
            for c in paren:
                if c == "(":
                    op += 1
                else:
                    if op == 0:
                        return False
                    else:
                        op -= 1
            return op == 0
        
        ans = []
        return [paren for paren in L if is_valid(paren)]

File: 22-generate-parentheses/generate-parentheses.cpp
class Solution {
public:
    void generate(int n, vector<string>& ans, string sofar, int op, int cl){
        if(op > n)
            return;
        if(cl > op)
            return;
        if(cl == n && op == n){
            ans.push_back(sofar);
        }
        generate(n, ans, sofar + "(", op + 1, cl);
        generate(n, ans, sofar + ")", op, cl + 1);
    }

    vector<string> generateParenthesis(int n) {
        vector<string> ans;
        generate(n, ans, "", 0, 0);
        return ans;
    }
};